crypto/ring-signature/src/ring_signature/mlsag_verify.rs

```diff
@@ -59,20 +59,6 @@ impl<'a, R: Ring> MlsagVerify<'a, R> {
         // point.
         self.ring.check()?;
 
-        // Scalars must be canonical.
-        let is_canonical = bool::from(self.c_zero.scalar.is_canonical());
-        if !is_canonical {
-            return Err(Error::InvalidCurveScalar);
-        }
-
-        // Scalars must be canonical.
-        for response in self.responses {
-            let is_canonical = bool::from(response.scalar.is_canonical());
-            if !is_canonical {
-                return Err(Error::InvalidCurveScalar);
-            }
-        }
-
```

Comment:
This is the only part of the PR that I am not super happy with, but the `is_canonical` method is no longer public. See https://github.com/dalek-cryptography/curve25519-dalek/commit/f460ae149b0000695205cc78f560d74a2d3918eb for details on why.
However, reviewing the code, it appears to me that there is no way in which these fields would break the invariant - since we're not using `from_bits` anywhere and the serde implementation uses the safe `from_canonical_bytes` (see https://github.com/dalek-cryptography/curve25519-dalek/blob/main/curve25519-dalek/src/scalar.rs#L451C38-L451C58)

Reply:
It does look like these fields hold the invariant. Looking at mslag it seems to only deal with CurveScalar and it looks like that is using from_bytes_mod_order() to ensure canonical.

It's unfortunate that some dependency crate could enable the "legacy_compatibility" feature and still create an invalid scalar using `from_bits`.

Reply:
Agreed, I suspect/hope that in the future that feature will eventually be removed.

Reply:
We could use a tool like semgrep to detect calls to `from_bits`. I think using `cargo vendor` it should be trivial to grab all the dependency code too.